### PR TITLE
fix copy module update atime/mtime

### PR DESCRIPTION
**Pull request by dkuji**
##### SUMMARY

- update atime, mtime

##### ISSUE TYPE

- Bugfix Pull Request
#83013

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: test/integration/targets/copy/tasks/tests.yml [hunk 1]
```diff
@@ -31,6 +31,22 @@
...
 
+- name: Check the stat results of the file
+  stat:
+    path: "{{ remote_file }}"
+  register: stat_results
+
+- name: Check the stat results of the src file
+  stat:
+    path: foo.txt
+  register: stat_results_src
```
Comment (flowerysong) [inline test/integration/targets/copy/tasks/tests.yml]: `foo.txt` is a local path, you can't use `stat` against the target to check it (at a minimum you need `delegate_to: localhost`. It's also not clear to me that this is actually a valid test case for this change; did you run it with the old code to verify that it fails?

**Review by mkrizek — COMMENTED** [review]
Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment).

File: test/integration/targets/copy/tasks/tests.yml (hunk 1 above)
Comment (s-hertel) [reply test/integration/targets/copy/tasks/tests.yml]: I would agree that this test case doesn't seem to cover the reported bug - it appears to be specific to `content`?

I can reproduce easily [like this](https://gist.github.com/s-hertel/0765d0d35ad1ae7be905a41e08805cf9) (and it doesn't appear to be specific to ubuntu focal or Python 3.10), but for some reason I cannot reproduce with ansible-test. Adding this to the start of the tests.yml file:

```yaml
- name: Create file with new content
  shell: 'echo "test" > {{ remote_tmp_dir }}/test_mtime'

- name: Stat remote file
  command: "stat {{ remote_tmp_dir }}/test_mtime"
  register: mtime1

- name: Test updating remote content
  copy:
    content: "test2"
    dest: "{{ remote_tmp_dir }}/test_mtime"

- name: Stat remote file
  command: "stat {{ remote_tmp_dir }}/test_mtime"
  register: mtime2

- debug: msg='{{ mtime1 }}'
- debug: msg='{{ mtime2 }}'
```
and running `ansible-test integration copy --docker` on devel results in an updated mtime:
```sh
TASK [copy : Create file with new content] *************************************
changed: [testhost]

TASK [copy : Stat remote file] *************************************************
changed: [testhost]

TASK [copy : Test updating remote content] *************************************
changed: [testhost]

TASK [copy : Stat remote file] *************************************************
changed: [testhost]

TASK [copy : debug] ************************************************************
ok: [testhost] => {
    "msg": {
        "changed": true,
        "cmd": [
            "stat",
            "/tmp/ansible.zcwpem54.test/test_mtime"
        ],
        "delta": "0:00:00.005443",
        "end": "2024-05-20 22:53:47.879092",
        "failed": false,
        "msg": "",
        "rc": 0,
        "start": "2024-05-20 22:53:47.873649",
        "stderr": "",
        "stderr_lines": [],
        "stdout": "  File: /tmp/ansible.zcwpem54.test/test_mtime\n  Size: 5         \tBlocks: 8          IO Block: 4096   regular file\nDevice: 8ch/140d\tInode: 106         Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2024-05-20 22:53:47.691753256 +0000\nModify: 2024-05-20 22:53:47.691753256 +0000\nChange: 2024-05-20 22:53:47.691753256 +0000\n Birth: 2024-05-20 22:53:47.691753256 +0000",
        "stdout_lines": [
            "  File: /tmp/ansible.zcwpem54.test/test_mtime",
            "  Size: 5         \tBlocks: 8          IO Block: 4096   regular file",
            "Device: 8ch/140d\tInode: 106         Links: 1",
            "Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)",
            "Access: 2024-05-20 22:53:47.691753256 +0000",
            "Modify: 2024-05-20 22:53:47.691753256 +0000",
            "Change: 2024-05-20 22:53:47.691753256 +0000",
            " Birth: 2024-05-20 22:53:47.691753256 +0000"
        ]
    }
}

TASK [copy : debug] ************************************************************
ok: [testhost] => {
    "msg": {
        "changed": true,
        "cmd": [
            "stat",
            "/tmp/ansible.zcwpem54.test/test_mtime"
        ],
        "delta": "0:00:00.002923",
        "end": "2024-05-20 22:53:48.408639",
        "failed": false,
        "msg": "",
        "rc": 0,
        "start": "2024-05-20 22:53:48.405716",
        "stderr": "",
        "stderr_lines": [],
        "stdout": "  File: /tmp/ansible.zcwpem54.test/test_mtime\n  Size: 5         \tBlocks: 8          IO Block: 4096   regular file\nDevice: 8ch/140d\tInode: 116         Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2024-05-20 22:53:48.231760031 +0000\nModify: 2024-05-20 22:53:48.231760031 +0000\nChange: 2024-05-20 22:53:48.231760031 +0000\n Birth: 2024-05-20 22:53:48.231760031 +0000",
        "stdout_lines": [
            "  File: /tmp/ansible.zcwpem54.test/test_mtime",
            "  Size: 5         \tBlocks: 8          IO Block: 4096   regular file",
            "Device: 8ch/140d\tInode: 116         Links: 1",
            "Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)",
            "Access: 2024-05-20 22:53:48.231760031 +0000",
            "Modify: 2024-05-20 22:53:48.231760031 +0000",
            "Change: 2024-05-20 22:53:48.231760031 +0000",
            " Birth: 2024-05-20 22:53:48.231760031 +0000"
        ]
    }
}
```
:confused:

Comment (dkuji) [reply test/integration/targets/copy/tasks/tests.yml]: As you say, this test did not cover the bug.

I tried to reproduce it with ansible-test integration.
I spent the last week without reproducing it, but I succeeded in reproducing it.

I found that the problem is reproduced when [Errno 18] Cross-device link: does not occur (when not crossing partitions) as a result of os.rename().

https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1616

This is why the problem is reproduced when dest is a relative path, but not when it is an absolute path such as "{{ remote_dir }}/foo.txt".

I modified test with dest as a relative path.

**Comment by dkuji** [comment]
@flowerysong @s-hertel @mkrizek PTAL?

**Comment by s-hertel** [comment]
@dkuji Sorry, this is still in my queue. I'm kind of concerned how easy it seems like it would be to accidentally lose the test coverage. I still haven't understood how to reliably reproduce the issue (for example, I was able to tweak your test to use the full path, and reproduce, so it's not limited to relative paths). Even though the test is destructive, I think the test should use the remote_tmp_dir if possible for easy cleanup.

The [`os.rename`](https://docs.python.org/3/library/os.html#os.rename) documentation mentions `The operation may fail on some Unix flavors if src and dst are on different filesystems.` - curious if that's related and would make a more intentional test.

File: test/integration/targets/copy/tasks/tests.yml [hunk 2]
```diff
@@ -2448,3 +2448,44 @@
       loop:
         - '{{ src }}'
         - '{{ src }}_dest'
+
+- name: Record the output directory
+  set_fact:
+    remote_file: foo.txt
```
Comment (bcoca) [inline test/integration/targets/copy/tasks/tests.yml]: A better way to do this a `block` with `vars:`

**Review by bcoca — APPROVED** [review]

File: test/integration/targets/copy/tasks/tests.yml (hunk 2 above)
Comment (s-hertel) [inline test/integration/targets/copy/tasks/tests.yml]: We could keep these files contained to the remote_tmp_dir.

```suggestion
    remote_file: "{{ remote_tmp_dir }}/foo.txt"
    ansible_remote_tmp: "{{ remote_tmp_dir }}"
```

**Review by s-hertel — APPROVED** [review]

**Review by s-hertel — APPROVED** [review]

File: test/integration/targets/copy/tasks/tests.yml (hunk 2 above)
Comment (s-hertel) [reply test/integration/targets/copy/tasks/tests.yml]: Updated

**Comment by grantma** [comment]
Checked the fix you put in for #83235. REferencing my comment in #83013, it does not fix the attempt to use shutil.copystat() via shutil.copy2() implemented further down in the os.rename() error recovery code in the atomic_move() function, like my patch set above does...

**Pull request merged**